src/ipc/ipc4/helper.c

```diff
@@ -453,32 +459,85 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 	if (!cpu_is_me(source->ipc_config.core) && !cross_core_bind)
 		return ipc4_process_on_core(source->ipc_config.core, false);
 
-	/* these might call comp_ipc4_get_attribute_remote() if necessary */
-	ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG, &source_src_cfg);
-	if (ret < 0) {
-		tr_err(&ipc_tr, "failed to get base config for module %#x", dev_comp_id(source));
-		return IPC4_FAILURE;
+	size = MODULE_MAX_SINKS * sizeof(struct ipc4_output_pin_format) +
+		MODULE_MAX_SOURCES * sizeof(struct ipc4_input_pin_format);
+
+	/* get obs from the base config extension if the src queue ID is non-zero */
+	if (bu->extension.r.src_queue) {
+		struct ipc4_output_pin_format out_fmt;
+		size_t output_fmt_offset;
+
+		src_basecfg_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  sizeof(*src_basecfg_ext) + size);
+		if (!src_basecfg_ext)
+			return IPC4_OUT_OF_MEMORY;
+
+		ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG_EXT, src_basecfg_ext);
+		if (!ret) {
+			size_t in_fmt_size = sizeof(struct ipc4_input_pin_format);
+			size_t out_fmt_size = sizeof(struct ipc4_output_pin_format);
+			output_fmt_offset =
```

Comment:
I think use uint8_t pin_formats[] is not suitable, it should be a structure? do you agree? as we don't know the exact size and data inside.

Reply:
This is sequence of 2 diffrent structures 
better to have a simple offset, anyway I agree that this is not easy to read :(

Reply:
yeah, some inline comments would help.